Add docs on difference between `OmegaConf` and `OmegaConfigLoader` (#3352)
kedro-org/kedro: docs/source/configuration/configuration_basics.md

`OmegaConfigLoader` can load `YAML` and `JSON` files. Acceptable file extensions are `.yml`, `.yaml`, and `.json`. By default, any configuration files used by the config loaders in Kedro are `.yml` files.

### `OmegaConf` vs. Kedro's `OmegaConfigLoader`

> `OmegaConf` is a configuration management library in Python that allows you to manage hierarchical configurations. On the other hand, Kedro's `OmegaConfigLoader` is a component within the Kedro framework that utilises `OmegaConf` for handling configurations.

Reviewer: ```suggestion
`OmegaConf` is a configuration management library in Python that allows you to manage hierarchical configurations. Kedro's `OmegaConfigLoader` uses `OmegaConf` for handling configurations.
```

> This means that when you work with `OmegaConfigLoader` in Kedro, you are leveraging the capabilities of `OmegaConf` without directly interacting with it.

Reviewer: ```suggestion
This means that when you work with `OmegaConfigLoader` in Kedro, you are using the capabilities of `OmegaConf` without directly interacting with it.
```

> `OmegaConfigLoader` in Kedro is designed to handle more complex configuration setups commonly used in Kedro projects. It automates the process of merging configuration files, such as those for catalogs, and takes into account different environments, making it convenient for managing configurations in a structured way.

Reviewer: ```suggestion
`OmegaConfigLoader` in Kedro is designed to handle more complex configuration setups commonly used in Kedro projects. It automates the process of merging configuration files, such as those for catalogs, and accounts for different environments to make it convenient to manage configurations in a structured way.
```

When you need to load configurations manually, such as for exploration in a notebook, you have two options:
1. Use the `OmegaConfigLoader` class provided by Kedro.
2. Directly use the `OmegaConf` library.

> If your use case involves loading only one configuration file and you don't have the complexity that Kedro's `OmegaConfigLoader` is designed to handle, it may be simpler to use `OmegaConf` directly.

Reviewer: ```suggestion
Kedro's `OmegaConfigLoader` is designed to handle complex project environments. If your use case involves loading only one configuration file and is straightforward, it may be simpler to use `OmegaConf` directly. 
```

```python
from omegaconf import OmegaConf

parameters = OmegaConf.load("/path/to/parameters.yml")
```

> When your configuration files are more complex and contain credentials or templating Kedro's `OmegaConfigLoader` is better suited to load configuration, as described in more detail in [How to load a data catalog with credentials in code?](#how-to-load-a-data-catalog-with-credentials-in-code) and [How to load a data catalog with templating in code?](advanced_configuration.md#how-to-load-a-data-catalog-with-templating-in-code).

Reviewer: ```suggestion
When your configuration files are complex and contain credentials or templating, Kedro's `OmegaConfigLoader` is more suitable, as described in more detail in [How to load a data catalog with credentials in code?](#how-to-load-a-data-catalog-with-credentials-in-code) and [How to load a data catalog with templating in code?](advanced_configuration.md#how-to-load-a-data-catalog-with-templating-in-code).
```

> In summary, while both `OmegaConf` and Kedro's `OmegaConfigLoader` provide ways to manage configurations, the latter is specifically tailored for Kedro projects with a focus on handling more intricate configuration structures and environments. The choice between them depends on the complexity of your configuration needs and whether you are working within the context of the Kedro framework.

Reviewer: ```suggestion
In summary, while both `OmegaConf` and Kedro's `OmegaConfigLoader` provide ways to manage configurations, your choice depends on the complexity of your configuration and whether you are working within the context of the Kedro framework. 
```

Reviewer: I've simplified this as I think you've already said it, but if you don't agree, feel free to reject this suggestion.

conf_catalog = conf_loader["catalog"]
```

### How to load a data catalog with credentials in code?

> Assuming your project contains a catalog and credentials file each located in a `base` and `local` environment respectively, you can use the `OmegaConfigLoader` to load these configurations and then pass them on to a `DataCatalog` object to get access to the catalog entries with resolved credentials.

Reviewer 2: In this section can we please put a `Note` admonition that explains why this shouldn't be done within a node, nodes should be pure functions with no knowledge of IO. Then mention hooks should be the only way to manipulate things like the catalog at runtime.

Author: I think that's a very good point and imo related to whether this is now in the right place in the docs. How do we distinguish between using "Kedro as a library" just in notebooks for exploration vs. as part of their Kedro projects where they should be following best practices? Any thoughts on this @stichbury and @astrojuanlu ?

Reviewer: Is the issue that people cannot use hooks in notebook exploration so have to put this in a node when they shouldn't? Because I'd think that is probably "OK" in as much as the code is not production-quality and it's not ideal but won't be world-ending. To prevent people doing this in nodes in production code projects when they shouldn't, a simple warning as Joel suggests should be sufficient, right? With a nod to the fact that "if you're unable to use hooks because you're doing exploratory notebook code" then we  see you, and don't forget to refactor your code when you move it into a project for best practice.

Reviewer 2: I'm working back from the many times we see users who do this inline within a node- my suggestion would be anytime we mention this pattern we should also say it's only to be used in situation X and not Y

Reviewer 2: I don't think there is any issue in repeating this in multiple places either

Reviewer 3: > Then mention hooks should be the only way to manipulate things like the catalog at runtime.

Hooks have a hard time manipulating the catalog at runtime https://github.com/kedro-org/kedro/issues/2728 better to leave that out of the note

> How do we distinguish between using "Kedro as a library" just in notebooks for exploration vs. as part of their Kedro projects where they should be following best practices?

I think it doesn't harm to include a reminder that node functions should indeed not do any I/O? This is not only a Kedro thing, but goes well beyond https://rhodesmill.org/brandon/slides/2015-05-pywaw/hoist/

Author: I've added a note, lmk if it's sufficient.

Reviewer 2: Love it! thanks @merelcht. I don't have approve / resolve rights anymore so take this as that!

conf_catalog = conf_loader["catalog"]
```

### How to load a data catalog with credentials in code?
```{note}

> It is not recommended to do load and manipulate a data catalog directly in a Kedro node. Nodes are designed to be pure functions and thus should remain agnostic of I/O.

Reviewer: ```suggestion
We do not recommend that you load and manipulate a data catalog directly in a Kedro node. Nodes are designed to be pure functions and thus should remain agnostic of I/O.
```

```

> Assuming your project contains a catalog and credentials file each located in a `base` and `local` environment respectively, you can use the `OmegaConfigLoader` to load these configurations and then pass them on to a `DataCatalog` object to get access to the catalog entries with resolved credentials.

Reviewer: ```suggestion
Assuming your project contains a catalog and credentials file, each located in `base` and `local` environments respectively, you can use the `OmegaConfigLoader` to load these configurations, and pass them to a `DataCatalog` object to access the catalog entries with resolved credentials.
```
